File: test-suite/src/test/java/io/micronaut/chatbots/telegram/UnknownCommandHandler.java
package io.micronaut.chatbots.telegram;

import io.micronaut.chatbots.core.SpaceParser;
import io.micronaut.chatbots.telegram.api.Chat;
import io.micronaut.chatbots.telegram.api.Update;
import io.micronaut.chatbots.telegram.api.send.SendMessage;
import io.micronaut.chatbots.telegram.core.SendMessageUtils;
import io.micronaut.chatbots.telegram.core.TelegramBotConfiguration;
import io.micronaut.chatbots.telegram.core.TelegramHandler;
import io.micronaut.context.annotation.Requires;
import io.micronaut.core.annotation.Nullable;
import io.micronaut.core.order.Ordered;
import jakarta.inject.Singleton;
import jakarta.validation.constraints.NotNull;

import java.util.Optional;

@Requires(property = "spec.name", value = "telegram")
// tag::class[]
@Singleton
class UnknownCommandHandler implements TelegramHandler<SendMessage> {

    private final SpaceParser<Update, Chat> spaceParser;

    UnknownCommandHandler(SpaceParser<Update, Chat> spaceParser) {
        this.spaceParser = spaceParser;
    }

    @Override
    public boolean canHandle(@Nullable TelegramBotConfiguration bot, @NotNull Update input) {
        return true; // <1>
    }

    @Override
    public Optional<SendMessage> handle(@Nullable TelegramBotConfiguration bot, @NotNull Update input) {
        return SendMessageUtils.compose(spaceParser, input, "I don't how to handle your query: %s".formatted(input.getMessage().getText()));
    }

    @Override
    public int getOrder() {
        return Ordered.LOWEST_PRECEDENCE; // <2>
    }
}
// end::class[]

File: test-suite-python/src/test/python/micronaut/chatbots/docs/telegram/UnknownCommandHandler.py
from jakarta.inject import Singleton
from java.util import Optional
from micronaut.chatbots.core import SpaceParser
from micronaut.chatbots.telegram.api import Chat, Update
from micronaut.chatbots.telegram.api.send import SendMessage
from micronaut.chatbots.telegram.core import SendMessageUtils, TelegramBotConfiguration, TelegramHandler
from micronaut.context.annotation import Executable, Requires
from micronaut.core.order import Ordered


@Requires(property="spec.name", value="telegram")
# tag::class[]
@Singleton
class UnknownCommandHandler(TelegramHandler[SendMessage]):

    def __init__(self, space_parser: SpaceParser[Update, Chat]):
        self.space_parser = space_parser

    def canHandle(self, bot: TelegramBotConfiguration | None, input: Update) -> bool:
        return True  # <1>

    def handle(self, bot: TelegramBotConfiguration | None, input: Update) -> Optional[SendMessage]:
        return SendMessageUtils.compose(self.space_parser, input, f"I don't know how to handle your query: {input.getMessage().getText()}")

    @Executable  # overrides the default getOrder() method of the Handler interface
    def getOrder(self) -> int:
        return Ordered.LOWEST_PRECEDENCE  # <2>
# end::class[]
